internal/datanode/compaction/compactor_common.go

```diff
@@ -33,29 +34,123 @@ import (
 	"github.com/milvus-io/milvus/pkg/log"
 	"github.com/milvus-io/milvus/pkg/util/tsoutil"
 	"github.com/milvus-io/milvus/pkg/util/typeutil"
+	"github.com/samber/lo"
 )
 
 const compactionBatchSize = 100
 
-func isExpiredEntity(ttl int64, now, ts typeutil.Timestamp) bool {
+type EntityFilter struct {
+	deletedPkTs map[interface{}]*tsHit // pk2ts
+	ttl         int64
+	currentTime time.Time
+
+	expiredCount int
+	deletedCount int
+}
+
+func newEntityFilter(deletedPkTs map[interface{}]*tsHit, ttl int64, currTime time.Time) *EntityFilter {
+	if deletedPkTs == nil {
+		deletedPkTs = make(map[interface{}]*tsHit)
+	}
+	return &EntityFilter{
+		deletedPkTs: deletedPkTs,
+		ttl:         ttl,
+		currentTime: currTime,
+	}
+}
+
+func (filter *EntityFilter) Filtered(pk any, ts typeutil.Timestamp) bool {
+	if filter.isEntityDeleted(pk, ts) {
+		filter.deletedCount++
+		return true
+	}
+
+	// Filtering expired entity
+	if filter.isEntityExpired(ts) {
+		filter.expiredCount++
+		return true
+	}
+	return false
+}
+
+func (filter *EntityFilter) GetExpiredCount() int {
+	return filter.expiredCount
+}
+
+func (filter *EntityFilter) GetDeletedCount() int {
+	return filter.deletedCount
+}
+
+func (filter *EntityFilter) GetDeltalogDeleteCount() int {
+	return len(filter.deletedPkTs)
+}
+
+func (filter *EntityFilter) GetMissingDeleteCount() int {
+	return len(lo.PickBy(filter.deletedPkTs, func(_ any, v *tsHit) bool {
+		return !v.hit
+	}))
+}
+
+func (filter *EntityFilter) isEntityDeleted(pk interface{}, ts typeutil.Timestamp) bool {
+	if pkTsHit, ok := filter.deletedPkTs[pk]; ok {
```

Comment:
IMO this is the only accurate calculation, because not hitted PK are really never been hitted. 

`deletedCount` might be greater than `delete entries` for dup PKs. Thus the metrics will be useless when meet dup pks, and very hard to know if it's usefully or not(because pk dup is hard to detect).

Maybe I can use a functional struct to simplify the pkTsHit map like 
```golang
type pkFilter struct {
    pk
    ts
    hitCount atomic.Int64
}
func (f *pkFilter) Filter(pk, ts) bool {}
func (f *pkFilter) Missing(){f.htCount.Load() == int64(0)}
```